CXSPA-3072: Rxjs 7 compatible updates

ercultimate opened the pull request.

ercultimate committed:
replace deprecated subscribe callback arguments with subscriber object

ercultimate committed:
update typing to clarify forkJoin arguments

ercultimate committed:
break down overstuffed pipe()

ercultimate committed:
provide then and else cases to iif

ercultimate committed:
fulfill Subject.next() type requirements

ercultimate committed:
always return an observable-like in defer

The 6.x behavior of returning undefined in defer results in an
observable that only completes, so a default of EMPTY is equivalent

ercultimate committed:
Subscription.add will not longer chain

ercultimate committed:
timer will no longer accept undefined

ercultimate committed:
fulfill next type by guarding against undefined token

ercultimate committed:
fix sonar issue where type argument was same as default

ercultimate committed:
add expected error handlers

ercultimate committed:
remove explicit type arguments from `of()`

plabadie commented on projects/core/src/auth/user-auth/services/auth-http-header.service.ts:
```diff
@@ -273,6 +273,7 @@ export class AuthHttpHeaderService implements OnDestroy {
           // we want to refresh the access token only when it is old.
           // this is a guard for the case when there are multiple parallel http calls
           if (
+            token &&
```
What is the purpose of adding `token &&` ?

ercultimate replied:
discussion results: This results in a logical change, so we will defer this to investigation for 7.0 release.  For now, we will do a type casting.

plabadie commented on projects/storefrontlib/cms-components/myaccount/my-coupons/my-coupons.component.ts:
```diff
@@ -93,21 +93,20 @@ export class MyCouponsComponent implements OnInit, OnDestroy {
+        .subscribe((error) => {
+          this.subscriptionFail(error);
+        })
+    );
```
Should the observer object be used here?

ercultimate replied:
No, only when using multiple callback functions is the listener object required.  When using a singular "next" callback function, we are not required to use a listener object.

plabadie commented on feature-libs/cart/base/core/facade/active-cart.service.ts:
```diff
@@ -378,54 +378,63 @@ export class ActiveCartService implements ActiveCartFacade, OnDestroy {
+        map(([cartState]) => cartState.value),
+        filter((cart): cart is Cart => !isEmpty(cart)),
+        take(1)
+      );
```
It's a bit hard to see the diff.  Can you help me understand the change? It looks like there are 2 new `pipe().` some new `?.` and `cart is Cart`.  ( all of which I presume must resolve an rxjs 7 issue)

ercultimate replied:
Correct, the pipe operator only supports up to 10 arguments before the typing gets too expensive to infer and falls back to `Observable<any>`. To keep the type information available, I broke the pipe arguments down into multiple chained calls with less than 10 arguments.  This is functionally equivalent to all pipe operators being in a single call.

The optional chaining operators are unnecessary and will be removed.

The observable type did not match the return type due to still including `undefined` in the type union.  I adjusted the `filter` function's return type to be a type guard to allow the `filter` to narrow the stream emit type.

ercultimate committed:
resolve code review comments

- remove uneccesary optional chain operators from active-cart service
- remove logical change in auth-http-header in favor of type casting

ercultimate committed:
Merge remote-tracking branch 'origin/epic/CXSPA-3078' into feature/CXSPA-3072
